src/containers/PipelinesDropdown/PipelinesDropdown.test.js

```diff
@@ -0,0 +1,254 @@
+      uid: 'id-pipeline-2'
+    }
+  },
+  'id-pipeline-3': {
+    metadata: {
+      name: 'pipeline-3',
+      namespace: 'green',
+      uid: 'id-pipeline-3'
+    }
+  }
+};
+
+const pipelinesStoreDefault = {
+  pipelines: {
+    byId: pipelinesById,
+    byNamespace: pipelinesByNamespace,
+    isFetching: false
+  }
+};
+
+const pipelinesStoreFetching = {
+  pipelines: {
+    byId: pipelinesById,
+    byNamespace: pipelinesByNamespace,
+    isFetching: true
+  }
+};
+
+const namespacesByName = {
+  blue: '',
+  green: ''
+};
+
+const namespacesStoreBlue = {
+  namespaces: {
+    byName: namespacesByName,
+    selected: 'blue'
+  }
+};
+
+const namespacesStoreGreen = {
+  namespaces: {
+    byName: namespacesByName,
+    selected: 'green'
+  }
+};
+
+const initialTextRegExp = new RegExp('select pipeline', 'i');
+
+const middleware = [thunk];
+const mockStore = configureStore(middleware);
+
+beforeEach(() => {
+  jest.spyOn(API, 'getPipelines').mockImplementation(() => pipelinesById);
+});
+
+it('PipelinesDropdown renders items based on Redux state', () => {
+  const store = mockStore({
+    ...pipelinesStoreDefault,
+    ...namespacesStoreBlue
+  });
+  const { getByText, queryByText } = render(
+    <Provider store={store}>
+      <PipelinesDropdown {...props} />
+    </Provider>
+  );
+  // View items
+  fireEvent.click(getByText(initialTextRegExp));
+  Object.keys(pipelinesByNamespace.blue).forEach(item => {
+    expect(queryByText(new RegExp(item, 'i'))).toBeTruthy();
+  });
+  // Select item
+  fireEvent.click(getByText(/pipeline-1/i));
+  Object.keys(pipelinesByNamespace.blue).forEach(item => {
+    if (item !== 'pipeline-1') {
+      expect(queryByText(new RegExp(item, 'i'))).toBeFalsy();
+    }
+  });
+  expect(queryByText(/pipeline-1/i)).toBeTruthy();
+  expect(queryByText(initialTextRegExp)).toBeFalsy();
+});
+
+it('PipelinesDropdown renders items based on Redux state when namespace changes', () => {
+  const blueStore = mockStore({
+    ...pipelinesStoreDefault,
+    ...namespacesStoreBlue
+  });
+  const { container, getByText, queryByText } = render(
+    <Provider store={blueStore}>
+      <PipelinesDropdown {...props} />
+    </Provider>
+  );
+  // View items
+  fireEvent.click(getByText(initialTextRegExp));
+  Object.keys(pipelinesByNamespace.blue).forEach(item => {
+    expect(queryByText(new RegExp(item, 'i'))).toBeTruthy();
+  });
+  fireEvent.click(getByText(initialTextRegExp));
+
+  // Change selected namespace from 'blue' to 'green'
+  const greenStore = mockStore({
+    ...pipelinesStoreDefault,
+    ...namespacesStoreGreen
+  });
+  render(
+    <Provider store={greenStore}>
+      <PipelinesDropdown {...props} />
+    </Provider>,
+    { container }
+  );
+  // View items
+  fireEvent.click(getByText(initialTextRegExp));
+  Object.keys(pipelinesByNamespace.green).forEach(item => {
+    expect(queryByText(new RegExp(item, 'i'))).toBeTruthy();
+  });
+  Object.keys(pipelinesByNamespace.blue).forEach(item => {
```

Comment:
This test doesn't preclude other items being present. What we should really do here is iterate through each of the items in the drop down and verify they exist in the store for the selected namespace.

This way we know all of the expected items are rendered, but nothing more can slip through.

Reply:
@AlanGreene good point. It is possible that there could be some random value that is being displayed in the Dropdown that is not in the store.

Do you know how I would be able to get all of the items in the dropdown? Would I have to query the DOM with something like `document.getElementById('pipelines-dropdown');` and dig through the child nodes, or is there a better way to get all the items that wouldn't lead to brittle tests?

Reply:
If you ensure each item used for testing contains a common substring, you could use `getAllByText(/insertSubstringHere/i)` to get all of the items

Reply:
Hm, so I understand why you would want to go about checking that all of the items in the Dropdown are in the store in the selected namespace. However, if I were to implement it like you say above, then my current code already covers this case because I check all of the possible namespaces in the store.

I understand your reasoning behind changing the the test though, so I will update it 👍